WIP - Test cases for base64 module

akash-suresh opened the pull request:
@Zac-HD I have made an attempt to add a few test cases for the `base64` module.

I had a doubt - This particular test case `test_b85_encode_with_padding_decode_round_trip`, is failing for `payload = b'\x00'`. I set `pad = True` while `b85encode`. The expectation is that `b85decode`, will implicitly remove the padding, but it fails to decode. Let me know your thoughts.
```   
 def test_b85_encode_with_padding_decode_round_trip(self, payload):
        x = base64.b85encode(payload, True)
        self.assertEqual(payload, base64.b85decode(x))
```

akash-suresh committed:
added basic test cases for base64

akash-suresh committed:
Added extra params to test_a85_encode_decode_round_trip

Zac-HD commented on tests/test_encode_decode.py:
```diff
@@ -1,12 +1,62 @@
 class TestBase64(unittest.TestCase):
-    # TODO: https://docs.python.org/3/library/base64.html
-    pass
+    @given(payload=st.binary())
```
As well as the `payload` argument, we could provide `altchars` (`st.none() | ...`) - in practice bugs are most common when we feed in allowed-but-unexpected combinations of options.

Same logic applies to passing `validate=st.booleans()`; it *shouldn't* make any difference, but it *might*.

akash-suresh replied:
I added `altchars` and `validate` arguments. 
From the docs for `altchars` - `Optional altchars should be a byte string of length 2 which specifies an alternative alphabet for the '+' and '/' characters.  This allows an application to e.g. generate url or filesystem safe Base64 strings.`

Based on this, I came up with this test.

```
@given(
        payload=st.binary(),
        altchars=(st.none() | st.binary(min_size=2, max_size=2)),
        validate=st.booleans(),
    )
    def test_b64_encode_decode_round_trip(self, payload, altchars, validate):
        x = base64.b64encode(payload, altchars=altchars)
        self.assertEqual(
            payload, base64.b64decode(x, altchars=altchars, validate=validate)
        )
```

But, I observe that this fails. Example - 
```
payload=b"\x00\x00\x00\x03\xd0",
altchars=b"\x009",
base64.b64encode(payload, altchars=altchars) = b"AAAAA9A="
```
So the `9` in the encoded string, gets decoded back to `/` (because of `altchars`)


I am not sure, if this is a bug. 

I looked up a lot of examples on the internet, and all of them use `altchars = b"_-"`

In that case the test case (as below), works completely fine.
```
@given(
        payload=st.binary(),
        altchars=(st.none() | st.just(b"_-")),
        validate=st.booleans(),
    )
    def test_b64_encode_decode_round_trip(self, payload, altchars, validate):
        x = base64.b64encode(payload, altchars=altchars)
        self.assertEqual(
            payload, base64.b64decode(x, altchars=altchars, validate=validate)
        )
```

How do we proceed?

akash-suresh replied:
How should I proceed on this?

Zac-HD commented on tests/test_encode_decode.py:
```diff
@@ -1,12 +1,62 @@
+        x = base64.urlsafe_b64encode(payload)
+        self.assertEqual(payload, base64.urlsafe_b64decode(x))
+
+    @given(payload=st.binary())
```
Could also vary `casefold` and `map01`, same logic as above.

akash-suresh replied:
Done!

Zac-HD commented on tests/test_encode_decode.py:
```diff
@@ -1,12 +1,62 @@
+        x = base64.b32encode(payload)
+        self.assertEqual(payload, base64.b32decode(x))
+
+    @given(payload=st.binary())
```
`casefold` again!

akash-suresh replied:
Done!

Zac-HD commented on tests/test_encode_decode.py:
```diff
@@ -1,12 +1,62 @@
+    @given(
+        payload=st.binary(),
+        foldspaces=st.booleans(),
+        wrapcol=st.integers(0, 10),
```
I'd write this as `st.just(0) | st.integers(0, 1000)` - so it's still `0` reasonably often, but also allows longer lines.

akash-suresh replied:
Done!

Zac-HD commented on tests/test_encode_decode.py:
```diff
@@ -1,12 +1,62 @@
+    @unittest.expectedFailure
+    @given(payload=st.binary())
+    def test_b85_encode_with_padding_decode_round_trip(self, payload):
+        x = base64.b85encode(payload, True)
```
Let's stick to keyword arguments - `b85encode(payload, pad=True)` is easier to interpret without checking the documentation!

What's happening here is that if the input has trailing null bytes *and is then padded with null bytes*, the trailing nulls in the message are removed along with the padding.

I'd suggest doing some experiments to see how this works (e.g. is it affected by whether the length is a multiple of four?  etc) and then using a `.filter(lambda payload: not payload.endswith(b"\x00"))` or whatever so that we just avoid trying cases which we know don't work.

Finally, please write up some detailed comments explaining what's happening and why we use such a complicated strategy!

akash-suresh replied:
- Added the keyword pad. 
- I did some experiments, and the issue doesn't occur when I make length a multiple of 4. It turns out, when length%4 != 0, it adds padding with `b"\x00"`, to make it a multiple of 4.

Zac-HD reviewed (commented):
Nice work @akash-suresh, this is a great start :grin: 

I think you've got the fundamentals down pat, so the comments below are about how we can take it to the next level.

akash-suresh commented:
Thanks @Zac-HD 😄. I shall look at the comments and make changes accordingly.

akash-suresh committed:
- Added arguments altchars, validate to b64encode & b64decode
- Added arguments casefold, map01 to b32decode
- Added arguments casefold to b16encode & b16decode
- Updated wrapcol argument strategy of a85encode to take 0 more often.
- Added argument pad to a85encode & a85decode
- Added argument pad to b85encode & b85decode
- Fixed usage of pad argument in a85decode & b85decode, by manually adding padding when len(payload)!=4

akash-suresh committed:
minor modification to add_padding function

akash-suresh commented on tests/test_encode_decode.py:
```diff
@@ -1,12 +1,86 @@
+    #     payload=st.binary(),
+    #     altchars=(st.none() | st.binary(min_size=2, max_size=2)),
+    #     validate=st.booleans(),
+    # )
```
Need to uncomment, and use the hypothesis strategy defined on top instead of the one below.

Zac-HD replied:
If you want to delete the commented-out code, I'd be happy to merge this using the strategies below :smile:

akash-suresh replied:
Hey @Zac-HD! I have removed the commented out code. 😄

Zac-HD commented on tests/test_encode_decode.py:
```diff
@@ -1,12 +1,86 @@
+    @given(payload=st.binary(), pad=st.booleans())
+    def test_b85_encode_decode_round_trip(self, payload, pad):
+        x = base64.b85encode(payload, pad=pad)
+        # adding padding manually to payload, when pad is True and len(payload)%4!=0
+        if pad:
+            payload = add_padding(payload)
```
Instead of adding the padding to match, let's document and skip over the cases were we *don't* expect our property to hold:

```python
from hypothesis import assume

# If the payload ends with a null byte and its length is not a multiple of four,
# null-padding the payload means that the original payload is unrecoverable.
# We therefore just skip over such test cases.
assume(len(payload) % 4 == 0 or not payload.endswith(b"\x00") or not pad)
```

akash-suresh replied:
@Zac-HD It fails for payload not ending with `b"\x00"` as well. 

```
Falsifying example: test_b85_encode_decode_round_trip(
    self=<test_encode_decode.TestBase64 testMethod=test_b85_encode_decode_round_trip>,
    payload=b'\x01',
    pad=True,
)


b'\x01\x00\x00\x00' != b'\x01'

Expected :b'\x01'
Actual   :b'\x01\x00\x00\x00'
```

Zac-HD replied:
Huh.  I think your earlier implementation was correct then, and we should go back to that!

akash-suresh replied:
Sure. 👍

akash-suresh committed:
Removed comment

Zac-HD commented:
Thanks @akash-suresh!  If you're still interested in writing `binascii` or `plistlib` tests, go for it :smile:

akash-suresh commented:
Thanks for merging @Zac-HD . This is my first open source contribution 👻 . I have already started on writing tests for `binascii`, will raise a PR by the end of today. 😄

Zac-HD commented:
Congratulations!

Zac-HD commented:
Cross-reference: https://github.com/HypothesisWorks/hypothesis/issues/2430
